cms/toolbar/toolbar.py

```diff
@@ -74,28 +76,31 @@ def structure_renderer(self):
 
     @cached_property
     def structure_mode_active(self):
-        structure = get_cms_setting('CMS_TOOLBAR_URL__BUILD')
-        return self.is_staff and structure in self.request.GET
+        try:
+            match = resolve(self.request.path_info)
```

Comment:
use request resolver match

Reply:
The resolver match is None here most of the time when set as a cached property. Removing the cached property allows resolver match to be used. When set as a cached property the only method of getting the information that resolver match would have is using this method.

Reply:
This is most likely because Djangos cache is set by the Middleware.

Reply:
Middleware should not trigger this property to be called.
This is why its a lazy object now

Reply:
The failures occur (50+) when the RequestFactory is used. I believe that this is due to it accessing the toolbar Middleware directly to mock generating a response. The request factory is used directly in some tests and via the get_page_request method in the test cases:

https://github.com/divio/django-cms/pull/6482/files#diff-2a8a0d85dd243bf778b381658688d623R452